### DISCUSSION TITLE
Route path giving 404 when root_path is provided and both root_path and path are given as same.

### DISCUSSION THREAD

**Discussion by faizanazim11** [fastapi/fastapi#11015] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.routing import APIRouter
from pydantic import BaseModel

class FastAPIConfig(BaseModel):
    title: str = "Route Test"
    description: str = "FastAPI App to test route bug"
    version: str = "0.1.0"
    root_path: str = "/star"
    docs_url: str = "/docs"
    redoc_url: str = "/redoc"
    openapi_url: str = "/openapi.json"


app = FastAPI(**FastAPIConfig().model_dump())
router = APIRouter(prefix="/star")

@router.get("/hello")
def hello_world():
    return "Hello World"

app.include_router(router)
```

### Description

* Deploy the above code behind any proxy like NGINX where proxy is _/star_.
* Try getting the result from path "{host}/star/star/hello".

Expected Result: String Hello World is returned with 200 status

Actual Result: 404 not found is returned.

Note: When the path is given as "{host}/star/star/star/hello" without changing the code, the route works which is incorrect.

### Operating System

Linux

### Operating System Details

The code was ran is Docker Container with base image as: python:3.11.4-slim-bookworm

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.11.4

### Additional Context

This issue is occuring only with fastapi 0.109.0 where the new starlette version is also used.

**Comment by Kludex** [comment]
Uvicorn version? Bump it to the latest please.

**Reply by faizanazim11** [reply]
Hi, I am using the latest version of Uvicorn as well. This issue was found with all latest packages getting used.

**Comment by iudeen** [comment]
Its not when given the same, the issue occurs even when some other `root_path` is set. 

```
from typing import Union

from fastapi import FastAPI, APIRouter

app = FastAPI()

app.root_path = "/api"

router = APIRouter(prefix="/api1")


@router.get("/hello")
def read_root():
    return {"Hello": "World"}

app.include_router(router)
```

```
curl -X 'GET' \
  'http://127.0.0.1/api/api1/hello' \
  -H 'accept: application/json'

{
  "detail": "Not Found"
}
```
```
curl -X 'GET' \
'http://127.0.0.1/api/api/api1/hello'

{
    "Hello": "World"
}
```

**Reply by iudeen** [reply]
Not able to trace out where the issue is, yet. It works fine in version 0.108. Issue is only with 0.109.

**Comment by iudeen** [comment]
I got what the issue is I guess..

this function:

https://github.com/encode/starlette/blob/433da65fc1431754dae0c3c87290b0421aa4a6f9/starlette/_utils.py#L96
```
def get_route_path(scope: Scope) -> str:
    root_path = scope.get("root_path", "")
    route_path = re.sub(r"^" + root_path, "", scope["path"])
    return route_path
```
When mount path matches the root path, the path partially gets removed because of the Regex.

For instance, when the request is `/api/api1/hello`, the output of the function is `1/hello`

This is what is happening. If this is an issue, it must be fixed in starlette.
![image](https://github.com/tiangolo/fastapi/assets/10519440/3a993a2a-ab01-4dd3-b963-d68bec3cf67e)

We may fix this by adding `\b` to regex query.
```
def get_route_path(scope: Scope) -> str:
    root_path = scope.get("root_path", "")
    route_path = re.sub(r"^" + root_path + r"\\b", "", scope["path"])
    return route_path
```

**Reply by faizanazim11** [reply]
The function re.sub also accepts the count of strings to be replaced.
I have tried modifying the function and used the same for testing, I tried adding a unit test in starlette with this scenario and ran the test. Surprisingly the test passed but the issue still persists.

Here is the code snippet I updated for the utility:

```python
def get_route_path(scope: Scope) -> str:
    root_path = scope.get("root_path", "")
    route_path = re.sub(r"^" + root_path, "", scope["path"], 1)
    return route_path
```

The test case I added:

```python
app = Starlette(routes=echo_paths_routes)
client = test_client_factory(
    app, base_url="https://www.example.org/", root_path="/sub"
)
response = client.get("/sub/sub/path")
assert response.status_code == 200
assert response.json() == {
    "name": "subpath",
    "path": "/sub/sub/path",
    "root_path": "/sub/sub",
}
```

Even more surprising part is even if the code is reset, the test case still passes.

The testcase reference is at function: _test_paths_with_root_path_:1306:1312 -> file: tests/test_routing.py

```python
response = client.get("/root/sub/path")
assert response.status_code == 200
assert response.json() == {
    "name": "subpath",
    "path": "/root/sub/path",
    "root_path": "/root/sub",
}
```

Input for get_route_path:

```json
{
  "type": "http",
  "http_version": "1.1",
  "method": "GET",
  "path": "/sub/sub/path",
  "raw_path": b"/sub/sub/path",
  "root_path": "/sub",
  "scheme": "https"
}
```

Output for route_path in case of updated code snippet and current code snippet is same:

```
/sub/path
```

**Comment by faizanazim11** [comment, marked as answer]
I found the solution, it seems uvicorn also has a root_path parameter. When launching fastapi using uvicorn and having a root_path with a proxy from fastapi version 0.109.0 onwards the root_path parameter in uvicorn is also required to solve this issue.

Old code for running fastapi from uvicorn:

```python
uvicorn.run("main:app", host="0.0.0.0", port=3000)
```

New code for running fastapi from uvicorn:

```python
uvicorn.run("main:app", host="0.0.0.0", port=3000, root_path="/api")
```

**Reply by fgsalomon** [reply]
This won't fix the issue when using FastAPI's TestClient in tests.
